# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Patch for WasmEdge 0.16.1 to fix C++20 compatibility
# This patch resolves conflicts between experimental/span.hpp's to_address()
# and C++20's std::to_address by conditionally using std::to_address when available.

--- a/include/experimental/span.hpp
+++ b/include/experimental/span.hpp
@@ -37,6 +37,8 @@ template <class T, class = void> struct defined_to_address : false_type {};
 template <class T>
 struct defined_to_address<T, void_t<decltype(pointer_traits<T>::to_address)>>
     : true_type {};
+
+#if !defined(__cpp_lib_to_address) || __cpp_lib_to_address < 201711L
 } // namespace detail
 
 template <class T> constexpr T *to_address(T *p) noexcept {
@@ -51,6 +53,13 @@ template <class T> constexpr auto to_address(const T &p) noexcept {
     return to_address(p.operator->());
   }
 }
+#else
+} // namespace detail
+
+// C++20 or later with std::to_address available - use it from std
+using std::to_address;
+
+#endif // !defined(__cpp_lib_to_address) || __cpp_lib_to_address < 201711L
 
 namespace detail {
 
